pkg/configauditreport/controller.go

```diff
@@ -0,0 +1,437 @@
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("computing spec hash: %w", err)
+		}
+
+		policiesHash, err := policies.Hash(string(resourceKind))
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("computing policies hash: %w", err)
+		}
+
+		log.V(1).Info("Checking whether configuration audit report exists")
+		hasReport, err := r.hasReport(ctx, resourceRef, resourceHash, policiesHash)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("checking whether configuration audit report exists: %w", err)
+		}
+
+		if hasReport {
+			log.V(1).Info("Configuration audit report exists")
+			return ctrl.Result{}, nil
+		}
+
+		reportData, err := r.evaluate(ctx, policies, resource)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("evaluating resource: %w", err)
+		}
+
+		reportBuilder := NewReportBuilder(r.Client.Scheme()).
+			Controller(resource).
+			ResourceSpecHash(resourceHash).
+			PluginConfigHash(policiesHash).
+			Data(reportData)
+		err = reportBuilder.Write(ctx, r.ReadWriter)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *ResourceController) hasReport(ctx context.Context, owner kube.ObjectRef, podSpecHash string, pluginConfigHash string) (bool, error) {
+	if kube.IsClusterScopedKind(string(owner.Kind)) {
+		return r.hasClusterReport(ctx, owner, podSpecHash, pluginConfigHash)
+	}
+	// TODO FindByOwner should accept optional label selector to further narrow down search results
+	report, err := r.ReadWriter.FindReportByOwner(ctx, owner)
+	if err != nil {
+		return false, err
+	}
+	if report != nil {
+		return report.Labels[starboard.LabelResourceSpecHash] == podSpecHash &&
+			report.Labels[starboard.LabelPluginConfigHash] == pluginConfigHash, nil
+	}
+	return false, nil
+}
+
+func (r *ResourceController) hasClusterReport(ctx context.Context, owner kube.ObjectRef, podSpecHash string, pluginConfigHash string) (bool, error) {
+	report, err := r.ReadWriter.FindClusterReportByOwner(ctx, owner)
+	if err != nil {
+		return false, err
+	}
+	if report != nil {
+		return report.Labels[starboard.LabelResourceSpecHash] == podSpecHash &&
+			report.Labels[starboard.LabelPluginConfigHash] == pluginConfigHash, nil
+	}
+	return false, nil
+}
+
+func (r *ResourceController) policies(ctx context.Context) (*policy.Policies, error) {
+	cm := &corev1.ConfigMap{}
+
+	err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: r.Config.Namespace,
+		Name:      "starboard-policies-config",
+	}, cm)
+	if err != nil {
+		return nil, err
+	}
+	return policy.NewPolicies(cm.Data), nil
+}
+
+func (r *ResourceController) evaluate(ctx context.Context, policies *policy.Policies, resource client.Object) (v1alpha1.ConfigAuditReportData, error) {
+	checks, err := policies.Eval(ctx, resource)
+	if err != nil {
+		return v1alpha1.ConfigAuditReportData{}, err
+	}
+
+	summary := v1alpha1.ConfigAuditSummary{}
+
+	for _, check := range checks {
+		if check.Success {
+			continue
+		}
+		switch check.Severity {
```

Comment:
I would export the severity switch  check to a util , is it commonly used , we have a similar code in :
https://github.com/aquasecurity/starboard/blob/main/pkg/plugin/polaris/plugin.go#L522

Reply:
I've extracted this logic to the `v1alpha1.ConfigAuditSummaryFromChecks(checks []v1alpha1.Check)` method.